Improve readme (#16)
randa-mu/blocklock-solidity: README.md

# blocklock-solidity

[![Solidity ^0.8.x](https://img.shields.io/badge/Solidity-%5E0.8.x-blue)](https://soliditylang.org/)
[![License: MIT](https://img.shields.io/badge/License-MIT-green)](LICENSE)
[![Foundry Tests](https://img.shields.io/badge/Tested%20with-Foundry-red)](https://book.getfoundry.sh/)

A Solidity library enabling on-chain blocklock encryption and decryption, from the [dcipher threshold network](https://dcipher.network/). This facilites secure time-based data unlocking mechanisms for smart contracts.

## ✨ Overview

> Controlling access to data based on time is crucial for various use cases, such as auctions, voting, and content release schedules. `blocklock-solidity` provides developers with tools to implement blocklock encryption on-chain, ensuring that encrypted data can only be decrypted after a specified block height, thus enhancing security and fairness in time-sensitive operations. 

Reviewer: We previously discussed being consistent in the docs and tutorials / blogs, in term of using timelock encryption or blocklock encryption. 
In this context, timelock encryption might be better for consistency.

Reviewer: That's fine for now, thanks

Author: I will update the readme to use timelock encryption. But keep in mind, this library is called blocklock.solidity which is slightly not consistent.

> This blocklock library is powered by the dciper threshold network using **BLS pairing-based signature scheme** and **identity-based encryption** to achieve data encryption toward a future block height without relying on a trusted third party.  It is especially useful in decentralized settings where there is no such trusted third party to enforce timing.

Reviewer 2: ```suggestion
This blocklock library is powered by the dcipher threshold network using **BLS pairing-based signature scheme** and **identity-based encryption** to achieve data encryption toward a future block height without relying on a trusted third party.  It is especially useful in decentralized settings where there is no such trusted third party to enforce timing.
```

> The library is designed with modularity and simplicity in mind, allowing developers to easily integrate it into their existing smart contract projects to achieve blocklock on-chain. Its extensible architecture makes it suitable for a wide range of applications that require robust on-chain randomness.

Reviewer 2: I dunno if 'achieving blocklock' is an established term (yet!). Are we assuming that devs know what blocklock means when they hit this repo?

Reviewer 2: >robust on-chain randomness.

this doesn't provide randomness, that's https://github.com/randa-mu/randomness-solidity

### Features

Powered by the dcipher threshold network and its threshold-based cryptographic schemes, this library offers:

> * **On-chain Blocklock Encryption**: Encrypt data that can only be decrypted after a specified block number.

Reviewer: Timelock Encryption

We will also support other condition types, other than blocknumbers.

Reviewer 2: agree with Tari here - blocklock is a lib name for timelock encryption

* **Decryption**: Implement custom logic that gets triggered when the decryption key is received, i.e., decryption of the Ciphertext.
* **Modular**: Supports pluggable signature schemes for verifying decryption keys, enabling flexible cryptographic backends.

## 🧩 Smart Contracts    

### Blocklock
Provides functionality to schedule encrypted data to be decrypted only after a certain block number.
* ✨ `AbstractBlocklockReceiver.sol` - An abstract contract that developers must extend to request blocklock encryption and receive decrypted data in their smart contracts.
* `BlocklockSender.sol` - Handles creation and tracking of blocklock encryption requests.
* `BlocklockSignatureScheme.sol` - The `BN254-BLS-BLOCKLOCK` signature scheme for validating messaging coming from the dcipher network. This scheme is registered via `SignatureSchemeAddressProvider.sol`.

### Decryption

> Since decryption keys must be securely verified based on the block condition, this library also includes contracts for requesting and processing decryption requests using a defined schema.

Author: My understanding based on the code in `fulfilDecryptionRequest()`: 

when the condition (block height) is met, the dcipher network will send decryption keys with a signature. Then the `DecryptionSender.sol` will need to verify the signature to ensure that it is valid from dcipher network before send the decryption key to decry the data. 

Do it make sense?

> Since decryption keys must be securely verified based on the block condition, this library also includes contracts for requesting and processing decryption requests using a defined schema.

Author: Maybe my English writing is not clear, I updated it with another version.

| DecryptionSender (Impl) | [0xea9111e44D23029945f2E46b2bFf26b04D15bd6F](https://calibration.filfox.info/en/address/0xea9111e44D23029945f2E46b2bFf26b04D15bd6F)   |
| SignatureSchemeAddressProvider | [0xD2b5084E68230D609AEaAe5E4cF7df9ebDd6375A](https://calibration.filfox.info/en/address/0xD2b5084E68230D609AEaAe5E4cF7df9ebDd6375A)   |
| BlocklockSignatureScheme | [0x62C9CF8Ff30177d8479eDaB017f38017bEbf10C2](https://calibration.filfox.info/en/address/0x62C9CF8Ff30177d8479eDaB017f38017bEbf10C2)   |

> 💡 **Note:** `BlocklockSender.sol` is the only contract developers need to interact with directly when creating timelock requests, as it abstracts the underlying implementation and upgradeability via proxy.

## ⚙️ How It Works

1. **Data Encryption (off-chain)**

     Generate the encrypted data (`TypesLib.Ciphertext`) with the dcipher threshold network public key for the decryption at specified block hight, using the [blocklock.js](https://github.com/randa-mu/blocklock-js). 

    The TypeScript library supports the following Solidity types: `uint256`, `int256`, `address`, `string`, `bool`, `bytes32`, `bytes`, `uint256[]`, `address[]`, and `struct`.
2. **Blocklock Request**

>     Interact with the on-chain contract at `blocklock.requestBlocklock()` to create a blocklock request. Submit the encrypted data and specify the chain height for decryption. After request is stored, and a `requestId` is generated for tracking.

Reviewer 3: ```suggestion
    Interact with the on-chain contract at `blocklock.requestBlocklock()` to create a blocklock request. Submit the encrypted data and specify the chain height for decryption. After your request is stored, a `requestId` is generated for tracking.
```

3. **Decryption**

>     Once the specified block number is reached, a callback function (`receiveBlocklock（）`) will be triggered to deliver the decryption key, allowing the unlocking of encrypted data.

Reviewer 3: @azixus @najienka @CluEleSsUK an important design decision we need to make clear here is _when_ the decryption key is provided and the callbacks are called. Is it at block N, at N+1 or at N+n for a small n depending on XYZ?

Reviewer: Currently it's at `N+1`.

## 🚀 Getting Started

### Installation

> To get started, install the **blocklock-solidity** & **blocklock-js** library in your smart contract project using your preferred development tool.

Reviewer 2: ```suggestion
To get started, install the **blocklock-solidity** & **blocklock-js** libraries in your smart contract project using your preferred development tool.
```

**Hardhat (npm)**

```sh
$ npm install blocklock-solidity
```

**Foundry** 
```sh
$ forge install randa-mu/blocklock-solidity
```

> Install **blocklock-js** so you can encrypt data in your applicatoin . 

Reviewer 3: ```suggestion
Install **blocklock-js** so you can encrypt data in your application. 
```

```sh
$ npm install blocklock-js
```

### How to use

#### 1. Import the library

Start by importing the `AbstractBlocklockReceiver.sol` abstract contract into your smart contract. This contract provides the interface for making blocklock requests and handling callbacks.

```solidity
// Import the Types library for managing ciphertexts
import {TypesLib} from "blocklock-solidity/src/libraries/TypesLib.sol";
// Import the AbstractBlocklockReceiver for handling blocklock decryption & callbacks
import {AbstractBlocklockReceiver} from "blocklock-solidity/src/AbstractBlocklockReceiver.sol";
```

#### 2. Extend the AbstractBlocklockReceiver contract
Your contract must inherit from `AbstractBlocklockReceiver` and initialize with the deployed `BlocklockSender (Proxy)` contract from your desired network in the constructor.

```solidity
contract MockBlocklockReceiver is AbstractBlocklockReceiver {
    constructor(address blocklockContract) AbstractBlocklockReceiver(blocklockContract) {}
    ...
}

> ```

Author: I did not delete it completely. Just putting it in the first step

#### 3. Request Blocklock encryption
Define a function to initiate blocklock encryption requests originating from your application.

> In this function, interact with the deployed `BlocklockSender` contract instance to register the encryption request on-chain. 

Reviewer 3: ```suggestion
In this function, interact with the deployed `BlocklockSender` contract instance to register the encryption request on-chain, as shown in the following example. 
```

The function should return a `requestId`, which can be stored within your contract for tracking and managing the lifecycle of the blocklock encryption request.

```solidity
function createBlocklockRequest(uint256 decryptionBlockNumber, TypesLib.Ciphertext calldata encryptedData)
        external
        returns (uint256)
    {
        // Create blocklock request
        requestId = blocklock.requestBlocklock(decryptionBlockNumber, encryptedData);
        // Store the Ciphertext
        encryptedValue = encryptedData;
        return requestId;
    }
```

#### 4. Handle the Blocklock Callback

> Once the blocklock request is registered, the dcipher network will monitor the blockchain and, upon reaching the specified block height, invoke the `receiveBlocklock()` callback function on your contract to deliver the decryption key.

Reviewer 3: ```suggestion
Once the blocklock request is registered, the dcipher network will monitor the blockchain and, upon reaching the specified block height, invoke the `receiveBlocklock()` callback function of your contract to deliver the decryption key.
```

        // Decrypt stored Ciphertext with the decryption key
        plainTextValue = abi.decode(blocklock.decrypt(encryptedValue, decryptionKey), (uint256));
    }
```
> 💡 **Note:** `blocklock.decrypt` automatically verifies the dcipher threshold decryption key for you because of the power of threshold signatures scheme!

#### 5. Deploy the `BlocklockHandler` contract

> Please check the supprted network section to ensure your desired network is supported before deployment. You also need to use the deployed **BlocklockSender (Proxy)** address to initialized your contract.

Reviewer 3: ```suggestion
Please check the supported network section to ensure your desired network is supported before deployment. You also need to use the deployed **BlocklockSender (Proxy)** address to initialize your contract.
```
I'd add a link to that section too.

> When deploying, make sure your desired network is supported. Use the deployed `BlocklockSender (Proxy)` address when initializing your contract.

Reviewer 3: This reads just like the previous line, no?

Example in Foundry Script for Filecoin Calibration: 
```solidity
address blocklockSenderProxy = 0xfF66908E1d7d23ff62791505b2eC120128918F44;
MockBlocklockReceiver mockBlocklockReceiver = new MockBlocklockReceiver(address(blocklockSenderProxy));
console.log("\nMockBlocklockReceiver deployed at: ", address(mockBlocklockReceiver));
```

#### 6. Off-chain data encoding and encryption
Use [blocklock.js](https://github.com/randa-mu/blocklock-js) to encode and encrypt data off-chain before calling the smart contract. It ensures data is encrypted using the dcipher threshold public key and specified block height.

Below is a sample JavaScript/TypeScript snippet:
```Javascript
// Encode the uint256 value
const encoder = new SolidityEncoder();
const msgBytes = encoder.encodeUint256(msg);

> const encodedMessage = getBytes(msgBytes);

Reviewer 3: is `getBytes` a built-in? I don't see it mentioned anywhere before

Reviewer: It's imported from `ethers.js`

As in:

`import { getBytes } from "ethers";`

@longfeiWan9  we could add it at the top of the JavaScript code block.
